rename interfaces in StreamInterface and StreamExecutorInterface to support both CUDA and ROCm

whchung opened the pull request:
In this pull request, 2 interface changes are proposed:

    1) StreamInterface::CudaStreamMemberHack()

    Despite the fact that StreamExecutor and GPU common runtime are largely
    orthogonal, this particular routine in StreamExecutor is used in GPU common
    runtime and a couple of other operators. In this commit it's renamed as
    StreamInterface::GpuStreamMemberHack() and their call sites are also changed.

    2) StreamExecutorInterface::CudaContextHack()

    This member is renamed to StramExecutorInterface::GpuContextHack().

    Changes introduced in this commit includes:

    - some StreamExecutor interfaces and CUDA implementation
    - GPU common runtime related to interface changes in StreamExecutor
    - operators affected by interface changes in StreamExecutor

whchung commented:
/cc @jlebar

jlebar commented on tensorflow/contrib/nccl/kernels/nccl_manager.cc:
```diff
@@ -445,7 +445,7 @@ void NcclManager::LoopKernelLaunches(NcclStream* nccl_stream) {
   ScopedActivateExecutorContext scoped_context(nccl_stream->executor);
   const cudaStream_t* cu_stream = reinterpret_cast<const cudaStream_t*>(
-      comm_stream->implementation()->CudaStreamMemberHack());
+      comm_stream->implementation()->GPUStreamMemberHack());
```
This change should be split out into a separate commit.

jlebar commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -3587,7 +3587,8 @@ bool CudnnSupport::DoPoolForward(
     const dnn::BatchDescriptor& output_dimensions,
-    DeviceMemory<double>* output_data) {
+    DeviceMemory<double>* output_data,
+    ScratchAllocator* workspace_allocator) {
```
This change (adding the new parameter) should be split out.

jlebar commented on tensorflow/stream_executor/rocm/rocm_activation.cc:
```diff
@@ -0,0 +1,42 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
```
Please update copyright dates.

jlebar commented on tensorflow/stream_executor/rocm/rocm_activation.h:
```diff
@@ -0,0 +1,60 @@
+  ~ScopedActivateExecutorContext();
+
+ private:
+
```
No space after `private:`.

Probably best if we figure out how to clang-format this change.

jlebar commented on tensorflow/stream_executor/rocm/rocm_diagnostics.h:
```diff
@@ -0,0 +1,97 @@
+  // Note: if we're running on a machine that has no GPUs, we don't want to
+  // produce very much log spew beyond saying, "looks like there's no ROCM
+  // kernel
+  // module running".
```
reflow

jlebar commented on tensorflow/stream_executor/rocm/rocm_helpers.h:
```diff
@@ -0,0 +1,104 @@
+
+// Converts a const DeviceMemory reference to its underlying typed pointer in
+// ROCM
+// device memory.
```
reflow

jlebar reviewed (commented):
Did a quick skim with some additional suggestions of places where this could be split up.

Splitting it up will also be helpful for checking it in: I suspect things are going to break in exciting ways when we try to check this in, so the smaller the pieces, the better.

jlebar commented:
BTW, this looks remarkably good to me on a first (very superficial) glance.

jlebar commented:
>> I suspect things are going to break in exciting ways when we try to check this in
> Actually I make sure nothing breaks on either CUDA or ROCm before I submit PRs.

I should have specified: What I was thinking about is checking this code in to "upstream" google.  This is  fundamental code used not only by TensorFlow, and we're probably going to have to modify some internal classes based on these API changes.  So smaller pieces are going to be a lot easier to land.

whchung committed:
[ROCm] Interface changes for StreamExecutor to support both CUDA and ROCm

1) StreamInterface::CudaStreamMemberHack()

Despite the fact that StreamExecutor and GPU common runtime are largely
orthogonal, this particular routine in StreamExecutor is used in GPU common
runtime and a couple of other operators. In this commit it's renamed as
StreamInterface::GpuStreamMemberHack() and their call sites are also changed.

2) StreamExecutorInterface::CudaContextHack()

This member is renamed to StramExecutorInterface::GpuContextHack().

Changes introduced in this commit includes:

- some StreamExecutor interfaces and CUDA implementation
- GPU common runtime related to interface changes in StreamExecutor
- operators affected by interface changes in StreamExecutor

whchung commented:
@jlebar I've updated this PR so it's minimal now (32 lines). Would you mind to take a look at it again? Thanks a lot. I'll continue file subsequent PRs to cover other parts of StreamExecutor for ROCm platform.

jlebar commented:
LGTM.  Would you mind updating the PR title to match the new description here?

chsigg commented:
Hi Wen-Heng. Would you maybe mind to change to change 'GPUStreamMemberHack' to 'GpuStreamMemberHack'? We generally capitalize acronyms as single words. Thanks!

whchung commented:
@chsigg Revised to `GpuStreamMemberHack` now

whchung commented:
ping. @jlebar / @chsigg

whchung commented:
@qlzh727 It seems the failing test has nothing to do with this PR. Would it be possible to force re-run the test?

qlzh727 commented:
Oh, its fine. We have changed our sync process, all the changes will be port to our internal system and test again there. The test here is just some early indicator before we port the change.

jlebar commented:
I'm in the process of merging this change internally; I just need a few more people to approve it.
